compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -538,15 +538,21 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
             }
             ExprKind::Adt(box AdtExpr {
                 adt_def,
-                variant_index: _,
+                variant_index,
                 args: _,
                 user_ty: _,
                 fields: _,
                 base: _,
-            }) => match self.tcx.layout_scalar_valid_range(adt_def.did()) {
-                (Bound::Unbounded, Bound::Unbounded) => {}
-                _ => self.requires_unsafe(expr.span, InitializingTypeWith),
-            },
+            }) => {
+                if adt_def.variant(variant_index).fields.iter().any(|x| x.safety == Safety::Unsafe)
```

Comment:
Could you add a `VariantDef::has_unsafe_fields` method and move the iteration logic there?

```suggestion
                if adt_def.variant(variant_index).has_unsafe_fields()
```